contracts/stake/src/contract.rs

```diff
@@ -180,23 +206,43 @@ impl StakingTrait for Staking {
 
     fn unbond(env: Env, sender: Address, stake_amount: i128, stake_timestamp: u64) {
         sender.require_auth();
-
         env.storage()
             .instance()
             .extend_ttl(INSTANCE_LIFETIME_THRESHOLD, INSTANCE_BUMP_AMOUNT);
 
         let config = get_config(&env);
 
-        let mut stakes = get_stakes(&env, &sender);
+        // check for rewards and withdraw them
+        let found_rewards: WithdrawableRewardsResponse =
+            Self::query_withdrawable_rewards(env.clone(), sender.clone());
+
+        if !found_rewards.rewards.is_empty() {
+            Self::withdraw_rewards(env.clone(), sender.clone());
+        }
 
+        for distribution_address in get_distributions(&env) {
+            let mut distribution = get_distribution(&env, &distribution_address);
+            let stakes = get_stakes(&env, &sender).total_stake;
+            let old_power = calc_power(&config, stakes, Decimal::one(), TOKEN_PER_POWER); // while bonding we use Decimal::one()
+            let new_power = calc_power(
+                &config,
+                stakes - stake_amount,
+                Decimal::one(),
+                TOKEN_PER_POWER,
+            );
+            update_rewards(
+                &env,
+                &sender,
+                &distribution_address,
+                &mut distribution,
+                old_power,
+                new_power,
+            );
+        }
+
+        let mut stakes = get_stakes(&env, &sender);
         remove_stake(&env, &mut stakes.stakes, stake_amount, stake_timestamp);
-        stakes.total_stake = stakes
-            .total_stake
-            .checked_sub(stake_amount)
-            .unwrap_or_else(|| {
-                log!(&env, "Stake: Unbond: underflow occured.");
-                panic_with_error!(&env, ContractError::ContractMathError);
-            });
+        stakes.total_stake -= stake_amount;
```

Comment:
This is something we need to make sure is done using safe math life before.
This diff is a great example.